File: log4j2-api/src/main/java/org/apache/logging/log4j/Logger.java
/*
* Licensed to the Apache Software Foundation (ASF) under one or more
* contributor license agreements.  See the NOTICE file distributed with
* this work for additional information regarding copyright ownership.
* The ASF licenses this file to You under the Apache License, Version 2.0
* (the "License"); you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
package org.apache.logging.log4j;

import org.apache.logging.log4j.message.Message;

/**
 * This is the central interface in the log4j package. Most logging
 * operations, except configuration, are done through this interface.
 * @doubt See LOG4J2-39.
 * @doubt See LOG4J2-16.
 */
public interface Logger {

  static Marker FLOW_MARKER = Marker.getMarker("FLOW");
  static Marker ENTRY_MARKER = Marker.getMarker("ENTRY", FLOW_MARKER);
  static Marker EXIT_MARKER = Marker.getMarker("EXIT", FLOW_MARKER);

  static Marker EXCEPTION_MARKER = Marker.getMarker("EXCEPTION");
  static Marker THROWING_MARKER = Marker.getMarker("THROWING", EXCEPTION_MARKER);
  static Marker CATCHING_MARKER = Marker.getMarker("CATCHING", EXCEPTION_MARKER);

  /**
   * Log entry to a method.
   */
  void entry();

  /**
   * Log entry to a method.
   * @param params The parameters to the method.
   * @doubt Use of varargs results in array creation which can be a substantial portion of no-op case.
   * LogMF/LogSF provides several overrides to avoid vararg except in edge cases. (RG) LogMF
   * and LogSF implement these in LogXF which calls logger.callAppenders. callAppenders is
   * part of the implementation and cannot be used by the API. Adding more methods here
   * and in AbstractLogger is sufficient.
   */
  void entry(Object... params);

  /**
   * Log exit from a method.
   */
  void exit();

  /**
   * Log exiting from a method with the result.
   * @param result The result being returned from the method call.
   */
  void exit(Object result);

  /**
   * Log an exception or error to be thrown.
   * @param t The Throwable.
   */
  void throwing(Throwable t);

  /**
   * Log an exception or error to be thrown.
   * @param level The logging Level.
   * @param t The Throwable.
   */
  void throwing(Level level, Throwable t);

  /**
   * Log an exception or error that has been caught.
   * @param t The Throwable.
   */
  void catching(Throwable t);

  /**
   * Log an exception or error that has been caught.
   * @param level The logging Level.
   * @param t The Throwable.
   */
  void catching(Level level, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#TRACE TRACE} level.
   *
   * @param message the message object to log.
   */
  void trace(String message);

  /**
   * Log a message at the <code>TRACE</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   * <p/>
   * <p>
   * See {@link #debug(String)} form for more detailed information.
   * </p>
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void trace(String message, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#TRACE TRACE} level.
   *
   * @param message the message object to log.
   */
  void trace(Object message);

  /**
   * Log a message at the <code>TRACE</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   * <p/>
   * <p>
   * See {@link #debug(String)} form for more detailed information.
   * </p>
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void trace(Object message, Throwable t);

  /**
   * Log a message with parameters at the <code>TRACE</code> level.
   * @param message the message to log.
   * @param params parameters to the message.
   */
  void trace(String message, Object... params);

  /**
   * Check whether this Logger is enabled for the TRACE  Level.
   *
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         TRACE, <code>false</code> otherwise.
   */
  boolean isTraceEnabled();

  /**
   * Check whether this Logger is enabled for the TRACE  Level.
   *
   * @param marker The marker data specific to this log statement.
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         TRACE, <code>false</code> otherwise.
   */
  boolean isTraceEnabled(Marker marker);

  /**
   * Log a message with the specific Marker at the TRACE level.
   *
   * @param msg the message string to be logged
   */
  void trace(Message msg);

  /**
   * Log a message with the specific Marker at the TRACE level.
   *
   * @param msg the message string to be logged
   * @param t   A Throwable or null.
   */
  void trace(Message msg, Throwable t);

  /**
   * Log a message with the specific Marker at the TRACE level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   */
  void trace(Marker marker, Message msg);

  /**
   * Log a message with the specific Marker at the TRACE level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   * @param t      A Throwable or null.
   */
  void trace(Marker marker, Message msg, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#DEBUG DEBUG} level.
   *
   * @param message the message object to log.
   */
  void debug(String message);

  /**
   * Log a message at the <code>DEBUG</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message to log.
   * @param t       the exception to log, including its stack trace.
   */
  void debug(String message, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#DEBUG DEBUG} level.
   *
   * @param message the message object to log.
   */
  void debug(Object message);

  /**
   * Log a message at the <code>DEBUG</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message to log.
   * @param t       the exception to log, including its stack trace.
   */
  void debug(Object message, Throwable t);

  /**
   * Log a message with parameters at the <code>DEBUG</code> level.
   * @param message the message to log.
   * @param params parameters to the message.
   */
  void debug(String message, Object... params);

  /**
   * Check whether this Logger is enabled for the DEBUG Level.
   *
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         DEBUG, <code>false</code> otherwise.
   */
  boolean isDebugEnabled();

  /**
   * Check whether this Logger is enabled for the DEBUG Level.
   *
   * @param marker The marker data specific to this log statement.
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         DEBUG, <code>false</code> otherwise.
   */
  boolean isDebugEnabled(Marker marker);

  /**
   * Log a message with the specific Marker at the DEBUG level.
   *
   * @param msg the message string to be logged
   */
  void debug(Message msg);

  /**
   * Log a message with the specific Marker at the DEBUG level.
   *
   * @param msg the message string to be logged
   * @param t   A Throwable or null.
   */
  void debug(Message msg, Throwable t);

  /**
   * Log a message with the specific Marker at the DEBUG level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   */
  void debug(Marker marker, Message msg);

  /**
   * Log a message with the specific Marker at the DEBUG level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   * @param t      A Throwable or null.
   */
  void debug(Marker marker, Message msg, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#INFO INFO} level.
   *
   * @param message the message object to log.
   */
  void info(String message);

  /**
   * Log a message at the <code>INFO</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void info(String message, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#INFO INFO} level.
   *
   * @param message the message object to log.
   */
  void info(Object message);

  /**
   * Log a message at the <code>INFO</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void info(Object message, Throwable t);

  /**
   * Log a message with parameters at the <code>INFO</code> level.
   * @param message the message to log.
   * @param params parameters to the message.
   * @doubt Likely to misinterpret existing log4j client code that intended to call info(Object,Throwable). Incurs array creation expense on every call.
   */
  void info(String message, Object... params);

  /**
   * Check whether this Logger is enabled for the INFO Level.
   *
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         INFO, <code>false</code> otherwise.
   */
  boolean isInfoEnabled();

  /**
   * Check whether this Logger is enabled for the INFO Level.
   *
   * @param marker The marker data specific to this log statement.
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         INFO, <code>false</code> otherwise.
   */
  boolean isInfoEnabled(Marker marker);

  /**
   * Log a message with the specific Marker at the TRACE level.
   *
   * @param msg the message string to be logged
   */
  void info(Message msg);

  /**
   * Log a message with the specific Marker at the INFO level.
   *
   * @param msg the message string to be logged
   * @param t   A Throwable or null.
   */
  void info(Message msg, Throwable t);

  /**
   * Log a message with the specific Marker at the INFO level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   */
  void info(Marker marker, Message msg);

  /**
   * Log a message with the specific Marker at the INFO level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   * @param t      A Throwable or null.
   */
  void info(Marker marker, Message msg, Throwable t);

 /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#WARN WARN} level.
   *
   * @param message the message object to log.
   */
  void warn(String message);

  /**
   * Log a message at the <code>WARN</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void warn(String message, Throwable t);

 /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#WARN WARN} level.
   *
   * @param message the message object to log.
   */
  void warn(Object message);

  /**
   * Log a message at the <code>WARN</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void warn(Object message, Throwable t);

  /**
   * Log a message with parameters at the <code>WARN</code> level.
   * @param message the message to log.
   * @param params parameters to the message.
   * @doubt Likely to misinterpret existing log4j client code that intended to call info(Object,Throwable). Incurs array creation expense on every call.
   */
  void warn(String message, Object... params);

  /**
   * Check whether this Logger is enabled for the WARN Level.
   *
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         WARN, <code>false</code> otherwise.
   */
  boolean isWarnEnabled();

  /**
   * Check whether this Logger is enabled for the WARN Level.
   *
   * @param marker The marker data specific to this log statement.
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         WARN, <code>false</code> otherwise.
   */
  boolean isWarnEnabled(Marker marker);

  /**
   * Log a message with the specific Marker at the WARN level.
   *
   * @param msg the message string to be logged
   */
  void warn(Message msg);

  /**
   * Log a message with the specific Marker at the WARN level.
   *
   * @param msg the message string to be logged
   * @param t   A Throwable or null.
   */
  void warn(Message msg, Throwable t);

  /**
   * Log a message with the specific Marker at the WARN level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   */
  void warn(Marker marker, Message msg);

  /**
   * Log a message with the specific Marker at the WARN level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   * @param t      A Throwable or null.
   */
  void warn(Marker marker, Message msg, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#ERROR ERROR} level.
   *
   * @param message the message object to log.
   */
  void error(String message);

  /**
   * Log a message at the <code>ERROR</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void error(String message, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#ERROR ERROR} level.
   *
   * @param message the message object to log.
   */
  void error(Object message);

  /**
   * Log a message at the <code>ERROR</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void error(Object message, Throwable t);

  /**
   * Log a message with parameters at the <code>ERROR</code> level.
   * @param message the message to log.
   * @param params parameters to the message.
   * @doubt Likely to misinterpret existing log4j client code that intended to call info(Object,Throwable). Incurs array creation expense on every call.
   */
  void error(String message, Object... params);

  /**
   * Check whether this Logger is enabled for the ERROR Level.
   *
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         ERROR, <code>false</code> otherwise.
   */
  boolean isErrorEnabled();

  /**
   * Check whether this Logger is enabled for the ERROR Level.
   *
   * @param marker The marker data specific to this log statement.
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         ERROR, <code>false</code> otherwise.
   */
  boolean isErrorEnabled(Marker marker);

  /**
   * Log a message with the specific Marker at the ERROR level.
   *
   * @param msg the message string to be logged
   */
  void error(Message msg);

  /**
   * Log a message with the specific Marker at the ERROR level.
   *
   * @param msg the message string to be logged
   * @param t   A Throwable or null.
   */
  void error(Message msg, Throwable t);

  /**
   * Log a message with the specific Marker at the ERROR level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   */
  void error(Marker marker, Message msg);

  /**
   * Log a message with the specific Marker at the ERROR level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   * @param t      A Throwable or null.
   */
  void error(Marker marker, Message msg, Throwable t);

  /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#FATAL FATAL} level.
   *
   * @param message the message object to log.
   */
  void fatal(String message);

  /**
   * Log a message at the <code>FATAL</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void fatal(String message, Throwable t);

    /**
   * Log a message object with the {@link org.apache.logging.log4j.Level#FATAL FATAL} level.
   *
   * @param message the message object to log.
   */
  void fatal(Object message);

  /**
   * Log a message at the <code>FATAL</code> level including the
   * stack trace of the {@link Throwable}<code>t</code> passed as parameter.
   *
   * @param message the message object to log.
   * @param t       the exception to log, including its stack trace.
   */
  void fatal(Object message, Throwable t);
  /**
   * Log a message with parameters at the <code>FATAL</code> level.
   * @param message the message to log.
   * @param params parameters to the message.
   * @doubt Likely to misinterpret existing log4j client code that intended to call info(Object,Throwable). Incurs array creation expense on every call.
   */
  void fatal(String message, Object... params);

  /**
   * Check whether this Logger is enabled for the FATAL Level.
   *
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         FATAL, <code>false</code> otherwise.
   */
  boolean isFatalEnabled();

  /**
   * Check whether this Logger is enabled for the FATAL Level.
   *
   * @param marker The marker data specific to this log statement.
   * @return boolean - <code>true</code> if this Logger is enabled for level
   *         FATAL, <code>false</code> otherwise.
   */
  boolean isFatalEnabled(Marker marker);

  /**
   * Log a message with the specific Marker at the FATAL level.
   *
   * @param msg the message string to be logged
   */
  void fatal(Message msg);

  /**
   * Log a message with the specific Marker at the FATAL level.
   *
   * @param msg the message string to be logged
   * @param t   A Throwable or null.
   */
  void fatal(Message msg, Throwable t);

  /**
   * Log a message with the specific Marker at the FATAL level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   */
  void fatal(Marker marker, Message msg);

  /**
   * Log a message with the specific Marker at the FATAL level.
   *
   * @param marker the marker data specific to this log statement
   * @param msg    the message string to be logged
   * @param t      A Throwable or null.
   */
  void fatal(Marker marker, Message msg, Throwable t);
}

File: log4j-api-scala_2.10/src/main/scala/org/apache/logging/log4j/scala/Logger.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache license, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the license for the specific language governing permissions and
 * limitations under the license.
 */
package org.apache.logging.log4j.scala

import org.apache.logging.log4j.message.{EntryMessage, Message, MessageFactory2}
import org.apache.logging.log4j.spi.ExtendedLogger
import org.apache.logging.log4j.{Level, Marker}

import scala.language.experimental.macros

class Logger(val delegate: ExtendedLogger) {

  private final val FQCN = classOf[Logger].getName

  
  def fatal(marker: Marker, message: Message): Unit =
  macro LoggerMacro.fatalMarkerMsg

  def fatal(marker: Marker, message: CharSequence): Unit =
  macro LoggerMacro.fatalMarkerCseq

  def fatal(marker: Marker, message: AnyRef): Unit =
  macro LoggerMacro.fatalMarkerObject

  def fatal(marker: Marker, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.fatalMarkerMsgThrowable

  def fatal(marker: Marker, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.fatalMarkerCseqThrowable

  def fatal(marker: Marker, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.fatalMarkerObjectThrowable

  def fatal(message: Message): Unit =
  macro LoggerMacro.fatalMsg

  def fatal(message: CharSequence): Unit =
  macro LoggerMacro.fatalCseq
  
  def fatal(message: AnyRef): Unit =
  macro LoggerMacro.fatalObject
  
  def fatal(message: Message, cause: Throwable): Unit =
  macro LoggerMacro.fatalMsgThrowable

  def fatal(message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.fatalCseqThrowable

  def fatal(message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.fatalObjectThrowable


  def error(marker: Marker, message: Message): Unit =
  macro LoggerMacro.errorMarkerMsg

  def error(marker: Marker, message: CharSequence): Unit =
  macro LoggerMacro.errorMarkerCseq

  def error(marker: Marker, message: AnyRef): Unit =
  macro LoggerMacro.errorMarkerObject

  def error(marker: Marker, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.errorMarkerMsgThrowable

  def error(marker: Marker, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.errorMarkerCseqThrowable

  def error(marker: Marker, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.errorMarkerObjectThrowable

  def error(message: Message): Unit =
  macro LoggerMacro.errorMsg

  def error(message: CharSequence): Unit =
  macro LoggerMacro.errorCseq

  def error(message: AnyRef): Unit =
  macro LoggerMacro.errorObject

  def error(message: Message, cause: Throwable): Unit =
  macro LoggerMacro.errorMsgThrowable

  def error(message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.errorCseqThrowable

  def error(message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.errorObjectThrowable


  def warn(marker: Marker, message: Message): Unit =
  macro LoggerMacro.warnMarkerMsg

  def warn(marker: Marker, message: CharSequence): Unit =
  macro LoggerMacro.warnMarkerCseq

  def warn(marker: Marker, message: AnyRef): Unit =
  macro LoggerMacro.warnMarkerObject

  def warn(marker: Marker, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.warnMarkerMsgThrowable

  def warn(marker: Marker, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.warnMarkerCseqThrowable

  def warn(marker: Marker, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.warnMarkerObjectThrowable

  def warn(message: Message): Unit =
  macro LoggerMacro.warnMsg

  def warn(message: CharSequence): Unit =
  macro LoggerMacro.warnCseq

  def warn(message: AnyRef): Unit =
  macro LoggerMacro.warnObject

  def warn(message: Message, cause: Throwable): Unit =
  macro LoggerMacro.warnMsgThrowable

  def warn(message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.warnCseqThrowable

  def warn(message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.warnObjectThrowable


  def info(marker: Marker, message: Message): Unit =
  macro LoggerMacro.infoMarkerMsg

  def info(marker: Marker, message: CharSequence): Unit =
  macro LoggerMacro.infoMarkerCseq

  def info(marker: Marker, message: AnyRef): Unit =
  macro LoggerMacro.infoMarkerObject

  def info(marker: Marker, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.infoMarkerMsgThrowable

  def info(marker: Marker, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.infoMarkerCseqThrowable

  def info(marker: Marker, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.infoMarkerObjectThrowable

  def info(message: Message): Unit =
  macro LoggerMacro.infoMsg

  def info(message: CharSequence): Unit =
  macro LoggerMacro.infoCseq

  def info(message: AnyRef): Unit =
  macro LoggerMacro.infoObject

  def info(message: Message, cause: Throwable): Unit =
  macro LoggerMacro.infoMsgThrowable

  def info(message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.infoCseqThrowable

  def info(message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.infoObjectThrowable


  def debug(marker: Marker, message: Message): Unit =
  macro LoggerMacro.debugMarkerMsg

  def debug(marker: Marker, message: CharSequence): Unit =
  macro LoggerMacro.debugMarkerCseq

  def debug(marker: Marker, message: AnyRef): Unit =
  macro LoggerMacro.debugMarkerObject

  def debug(marker: Marker, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.debugMarkerMsgThrowable

  def debug(marker: Marker, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.debugMarkerCseqThrowable

  def debug(marker: Marker, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.debugMarkerObjectThrowable

  def debug(message: Message): Unit =
  macro LoggerMacro.debugMsg

  def debug(message: CharSequence): Unit =
  macro LoggerMacro.debugCseq

  def debug(message: AnyRef): Unit =
  macro LoggerMacro.debugObject

  def debug(message: Message, cause: Throwable): Unit =
  macro LoggerMacro.debugMsgThrowable

  def debug(message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.debugCseqThrowable

  def debug(message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.debugObjectThrowable


  def trace(marker: Marker, message: Message): Unit =
  macro LoggerMacro.traceMarkerMsg

  def trace(marker: Marker, message: CharSequence): Unit =
  macro LoggerMacro.traceMarkerCseq

  def trace(marker: Marker, message: AnyRef): Unit =
  macro LoggerMacro.traceMarkerObject

  def trace(marker: Marker, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.traceMarkerMsgThrowable

  def trace(marker: Marker, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.traceMarkerCseqThrowable

  def trace(marker: Marker, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.traceMarkerObjectThrowable

  def trace(message: Message): Unit =
  macro LoggerMacro.traceMsg

  def trace(message: CharSequence): Unit =
  macro LoggerMacro.traceCseq

  def trace(message: AnyRef): Unit =
  macro LoggerMacro.traceObject

  def trace(message: Message, cause: Throwable): Unit =
  macro LoggerMacro.traceMsgThrowable

  def trace(message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.traceCseqThrowable

  def trace(message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.traceObjectThrowable
  
  
  def log(level: Level, marker: Marker, message: Message): Unit =
  macro LoggerMacro.logMarkerMsg

  def log(level: Level, marker: Marker, message: CharSequence): Unit =
  macro LoggerMacro.logMarkerCseq

  def log(level: Level, marker: Marker, message: AnyRef): Unit =
  macro LoggerMacro.logMarkerObject

  def log(level: Level, marker: Marker, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.logMarkerMsgThrowable

  def log(level: Level, marker: Marker, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.logMarkerCseqThrowable
  
  def log(level: Level, marker: Marker, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.logMarkerObjectThrowable
  
  def log(level: Level, message: Message): Unit =
  macro LoggerMacro.logMsg

  def log(level: Level, message: CharSequence): Unit =
  macro LoggerMacro.logCseq

  def log(level: Level, message: AnyRef): Unit =
  macro LoggerMacro.logObject
  
  def log(level: Level, message: Message, cause: Throwable): Unit =
  macro LoggerMacro.logMsgThrowable

  def log(level: Level, message: CharSequence, cause: Throwable): Unit =
  macro LoggerMacro.logCseqThrowable

  def log(level: Level, message: AnyRef, cause: Throwable): Unit =
  macro LoggerMacro.logObjectThrowable
  

  def isEnabled(level: Level): Boolean = delegate.isEnabled(level)

  def isEnabled(level: Level, marker: Marker): Boolean = delegate.isEnabled(level, marker)


  def traceEntry(): EntryMessage = delegate.traceEntry()

  def traceEntry(params: AnyRef*): EntryMessage =
    delegate.traceEntry(null, params) // TODO would be useful to have this as a macro to avoid varargs array creation

  def traceEntry(message: Message): EntryMessage = delegate.traceEntry(message)

  def traceExit[R](result: R): R = delegate.traceExit(result)

  def traceExit(): Unit = delegate.traceExit()

  def traceExit(entryMessage: EntryMessage): Unit = delegate.traceExit(entryMessage)

  def throwing[T <: Throwable](t: T): T = delegate.throwing(t)

  def throwing[T <: Throwable](level: Level, t: T): T = delegate.throwing(level, t)

  def catching(t: Throwable): Unit = delegate.catching(t)

  def catching(level: Level, t: Throwable): Unit = delegate.catching(level, t)


  def level: Level = delegate.getLevel

  def name: String = delegate.getName

  def messageFactory: MessageFactory2 = delegate.getMessageFactory.asInstanceOf[MessageFactory2]

  /** Always logs a message at the specified level. It is the responsibility of the caller to ensure the specified
    * level is enabled.
    *
    * @param level    log level
    * @param marker   marker or `null`
    * @param message  message
    * @param cause    cause or `null`
    */
  private[scala] def logMessage(level: Level, marker: Marker, message: Message, cause: Throwable): Unit = {
    delegate.logMessage(FQCN, level, marker, message, cause)
  }

  /** Always logs a message at the specified level. It is the responsibility of the caller to ensure the specified
    * level is enabled.
    *
    * @param level    log level
    * @param marker   marker or `null`
    * @param message  message
    * @param cause    cause or `null`
    */
  private[scala] def logMessage(level: Level, marker: Marker, message: CharSequence, cause: Throwable): Unit = {
    delegate.logMessage(FQCN, level, marker, messageFactory.newMessage(message), cause)
  }

  /** Always logs a message at the specified level. It is the responsibility of the caller to ensure the specified
    * level is enabled.
    *
    * @param level    log level
    * @param marker   marker or `null`
    * @param message  message
    * @param cause    cause or `null`
    */
  private[scala] def logMessage(level: Level, marker: Marker, message: AnyRef, cause: Throwable): Unit = {
    delegate.logMessage(FQCN, level, marker, messageFactory.newMessage(message), cause)
  }

}
